### PR TITLE
[16.0][ADD] partner_middlename

### PR DESCRIPTION
**Pull request by Nikul-OSI**
@dreispt Can you please take a look?

### CODE REVIEW

**Comment by dreispt** [comment]
I think we can make this simpler - the idea would be to not add config options, and just have "First + Middle" where we would have "First" name.
Will add more specific suggestions.

File: partner_middlename/models/res_config_settings.py [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+class ResConfigSettings(models.TransientModel):
+    _inherit = "res.config.settings"
+
+    def _partner_names_order_selection(self):
+        options = super()._partner_names_order_selection()
+        new_labels = {
+            "last_first": "Firstname Middlename Lastname",
+            "last_first_comma": "Lastname, Firstname Middlename",
+        }
+        return [(k, new_labels[k]) if k in new_labels else (k, v) for k, v in options]
```
Comment (dreispt) [inline partner_middlename/models/res_config_settings.py]: Remove. We will just always add Middle name after First name.

File: partner_middlename/models/res_partner.py [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+from odoo.addons.partner_firstname import exceptions
+
+
+class ResPartner(models.Model):
+    _inherit = "res.partner"
+
+    middlename = fields.Char("Middle Name")
+
+    @api.model
+    def _get_computed_name(self, lastname, firstname, middlename=None):
```
Comment (dreispt) [inline partner_middlename/models/res_partner.py:15]: This method would be better refactored in the `partner_firstname` module, adding a helper prepare method, but that might be too much for today.

File: partner_middlename/models/res_partner.py [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+        """Compute the name combined with the second lastname too.
+
+        We have 2 lastnames, so lastnames and firstname will be separated by a
```
Comment (dreispt) [inline partner_middlename/models/res_partner.py]: Please revise

File: partner_middlename/models/res_partner.py [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+        else:
+            if lastname:
+                names.append(lastname)
+            if names and firstname and order == "last_first_comma":
+                names[-1] = names[-1] + ","
+            if firstname:
+                names.append(firstname)
+            if middlename:
+                names.append(middlename)
+        return " ".join(names)
```
Comment (dreispt) [inline partner_middlename/models/res_partner.py]: Please add tests.

File: partner_middlename/models/res_partner.py [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+    def _compute_name(self):
+        """Write :attr:`~.name` according to splitted data."""
+        for partner in self:
+            partner.name = self._get_computed_name(
+                partner.lastname,
+                partner.firstname,
+                partner.middlename,
+            )
+
+    def _inverse_name(self):
```
Comment (dreispt) [inline partner_middlename/models/res_partner.py:46]: The inverse was not working on my tests - setting the Name should compute the First+Middle+Last.
Can you check please?

**Review by dreispt — CHANGES_REQUESTED** [review]

**Comment by Nikul-OSI** [comment]
Hello @dreispt Code updated base on your comments, So can you please take a look?

File: partner_middlename/tests/test_config.py [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+        self.wizard.partner_names_order = "last_first"
+        self.wizard.set_values()
+        self.wizard.action_recalculate_partners_name()
+        self.assertEqual(self.partner.name, "Last First Middle")
+
+    def test_last_first_comma(self):
+        self.wizard.partner_names_order = "last_first_comma"
+        self.wizard.set_values()
+        self.wizard.action_recalculate_partners_name()
+        self.assertEqual(self.partner.name, "Last, First, Middle")
```
Comment (dreispt) [inline partner_middlename/tests/test_config.py]:
```suggestion
        self.assertEqual(self.partner.name, "Last, First Middle")
```

File: partner_middlename/tests/test_config.py [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+
+    def test_first_last(self):
+        self.wizard.partner_names_order = "first_last"
+        self.wizard.set_values()
+        self.wizard.action_recalculate_partners_name()
+        self.assertEqual(self.partner.name, "First Middle Last")
```
Comment (dreispt) [inline partner_middlename/tests/test_config.py:32]: Can you add a test for the reverse rule: set the name to "First Middle Last", or ""Last, First Middle", and see the idivisual fields properly populated.
Thanks.

File: partner_middlename/tests/test_name.py [hunk 8]
```diff
@@ -0,0 +1,176 @@
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl.html).
+# Copyright 2015 Grupo ESOC Ingeniería de Servicios, S.L.U.
+# Copyright 2015 Antiun Ingenieria S.L. - Antonio Espinosa
+
+from odoo.tests.common import TransactionCase
+
+from odoo.addons.partner_firstname.tests.base import MailInstalled
+
+
+class CompanyCase(TransactionCase):
```
Comment (dreispt) [inline partner_middlename/tests/test_name.py]: I believe this is the same test as in "partner_firstname", is redundant and can be removed.

File: partner_middlename/README.rst [hunk 9]
```diff
@@ -0,0 +1,119 @@
...
+   :alt: Odoo Community Association
+   :target: https://odoo-community.org
+
+OCA, or the Odoo Community Association, is a nonprofit organization whose
+mission is to support the collaborative development of Odoo features and
+promote its widespread use.
+
+This module is part of the `OCA/partner-contact <https://github.com/OCA/partner-contact/tree/16.0/partner_second_lastname>`_ project on GitHub.
+
+You are welcome to contribute. To learn how please visit https://odoo-community.org/page/Contribute.
```
Comment (dreispt) [inline partner_middlename/README.rst]: Part of this are still from the copied module you used as a starting point.
In that case it is better to replace the full content with a line such as "AUTOMATICALLY GENERATED".

**Review by dreispt — CHANGES_REQUESTED** [review]

File: partner_middlename/README.rst (hunk 9 above)
Comment (Nikul-OSI) [reply partner_middlename/README.rst]: @dreispt Yes code updated

File: partner_middlename/README.rst [hunk 10]
```diff
@@ -0,0 +1,119 @@
+========================
+Partner second last name
+========================
+
```
Comment (diogocsc) [inline partner_middlename/README.rst]: @Nikul-OSI please update this file to say "this will be Autogenerated" or similar

**Review by diogocsc — APPROVED** [review]

**Comment by diogocsc** [comment]
@Nikul-OSI is this PR still WIP? If not, please remove the [WIP] from its name

**Comment by Nikul-OSI** [comment]
> @Nikul-OSI is this PR still WIP? If not, please remove the [WIP] from its name

@diogocsc Yes Did, Ready for review.

**Review by quannguyen1560 — APPROVED** [review]

**Review by dreispt — APPROVED** [review]

**Pull request merged**